src/expressions/operators/compares/Compare.ts

```diff
@@ -134,27 +152,39 @@ class Compare extends Expression {
 				if (firstSequence.isEmpty() || secondSequence.isEmpty()) {
 					return sequenceFactory.empty();
 				}
-				// Node compares should not be atomized
-				return this._evaluationFunction(firstSequence, secondSequence, dynamicContext)
-					? sequenceFactory.singletonTrueSequence()
-					: sequenceFactory.singletonFalseSequence();
+
+				if (
+					firstSequence.first().type === this._firstType &&
+					secondSequence.first().type === this._secondType
+				) {
+					// Node compares should not be atomized
+					return this._evaluationFunction(firstSequence, secondSequence, dynamicContext)
+						? sequenceFactory.singletonTrueSequence()
+						: sequenceFactory.singletonFalseSequence();
+				}
 			} else {
-				const firstAtomizedSequence = atomize(firstSequence, executionParameters);
-				const secondAtomizedSequence = atomize(secondSequence, executionParameters);
+				firstAtomizedSequence = atomize(firstSequence, executionParameters);
```

Comment:
The one problem I see with that is that the types have to be as specific as possible since this check is in place to make sure the `_evaluationFunction` is actually correct in the case where things get annotated as attributes.

Reply:
We may do atomization in a lot of places. I think it might make sense to add a new function that returns the atomized type of something that is being put in, so that we can reuse that in a lot of places. Something like `computeTypeAfterAtomizing`? What would be your thoughts @HectorPeeters

Reply:
But doesn't fetching the atomized type just come down to calling `atomizedSequence.first().type`. Or should that function also atomize the sequence itself?